sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -199,23 +199,23 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
     // TimestampWritable.floatToTimestamp
     case FloatType =>
       buildCast[Float](_, f => try {
-        decimalToTimestamp(Decimal(f))
+        decimalToTimestamp(Decimal(f.toString))
```

Comment:
+1

as a follow, we might investigate why we are going (float -> decimal -> timestamp) here instead of skipping the decimal step.